firebase-components/src/main/java/com/google/firebase/inject/Deferred.java

```diff
@@ -37,11 +37,11 @@
  * }
  * }</pre>
  */
-public interface Deferred<T> {
+public interface Deferred<T> extends Provider<T> {
   /** Used by dependers to register their callbacks. */
   interface DeferredHandler<T> {
     @DeferredApi
-    void handle(Provider<T> provider);
+    void handle(T instance);
```

Comment:
I would like to avoid this, since for one it's a breaking change and also is less general and breaks laziness(if  eagerness is not required by other products).

What is this achieving other than removing the need to call `get()` once?

Reply:
The main advantage is indeed the removal of the need to call `get()`: instead of giving you an indirect reference to the data it gives you the data directly. When I used this code I first thought "why isn't it just giving me the instance? why is it giving me a provider to the instance?" I see your points about lazy initialization though.

What if the argument to `handle()` is changed from `Provider<T>` to `Deferred<T>` and the `this` instance is passed as the argument? Since `Deferred` now extends `Provider` this provides one authoritative way to get the instance: call `get()` on the `Deferred` object.

Also, since this is an internal-only API (and I don't see any products using `Deferred` yet), could you explain the concerns about this being a breaking change?

Reply:
Note: The discussion about modifying `Deferred` to extend `Provider` occurred here: https://github.com/firebase/firebase-android-sdk/pull/2418#pullrequestreview-586863949

Reply:
I feel like `Deferred` and `Provider` are completely disjoint concepts - even though related in the context of dependency injection. The fact that `OptionalProvider` implements both is just an internal implementation detail that I would prefer not to leak into the interface.

> What if the argument to handle() is changed from Provider<T> to Deferred<T> and the this instance is passed as the argument? Since Deferred now extends Provider this provides one authoritative way to get the instance: call get() on the Deferred object.

I guess we could do that, but I don't see the benefits of this approach, afaict it's completely isomorphic to the existing interface however:

* has the downside I described above
* provides 2 ways of doing the same thing, i.e. calling get on the deferred itself or on the provider passed into `handle()`
* constrains the implementation of deferred as its contract becomes more complicated.

Looks like rtdb's change has no problem with deferred not extending provider and instead uses an atomic reference to determine if auth is present or not, can we do the same in firestore?

Reply:
I've reverted the changes to `Deferred` and `OptionalProvider`. It is indeed possible to use `Deferred` without the changes I suggested and I've made those changes as well.